File: src/games/stendhal/server/events/SoundEvent.java
/* $Id$ */
/***************************************************************************
 *                   (C) Copyright 2003-2014 - Stendhal                    *
 ***************************************************************************
 ***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/
package games.stendhal.server.events;

import games.stendhal.common.constants.Events;
import games.stendhal.common.constants.SoundID;
import games.stendhal.common.constants.SoundLayer;
import marauroa.common.game.Definition.DefinitionClass;
import marauroa.common.game.Definition.Type;
import marauroa.common.game.RPClass;
import marauroa.common.game.RPEvent;

/**
 * A sound.
 *
 * @author hendrik
 */
public class SoundEvent extends RPEvent {

	/**
	 * Creates the rpclass.
	 */
	public static void generateRPClass() {
		final RPClass rpclass = new RPClass(Events.SOUND);
		rpclass.add(DefinitionClass.ATTRIBUTE, "sound", Type.STRING);
		rpclass.add(DefinitionClass.ATTRIBUTE, "sound_id", Type.STRING);
		rpclass.add(DefinitionClass.ATTRIBUTE, "radius", Type.INT);
		rpclass.add(DefinitionClass.ATTRIBUTE, "volume", Type.INT);
		rpclass.add(DefinitionClass.ATTRIBUTE, "layer", Type.BYTE);
	}

	/**
	 * Creates a new sound event with an infinite range.
	 *
	 * @param sound
	 * 		Name of sound to play.
	 * @param layer
	 * 		Sound layer (e. g. ambient sound).
	 */
	public SoundEvent(final String sound, final SoundLayer layer) {
		super(Events.SOUND);
		put("sound", sound);
		put("layer", layer.ordinal());
	}

	/**
	 * Creates a new sound event using an ID with an infinite range.
	 *
	 * @param id
	 * 		SoundID to send with event.
	 * @param layer
	 * 		Sound layer (e. g. ambient sound).
	 */
	public SoundEvent(final SoundID id, final SoundLayer layer) {
		super(Events.SOUND);
		put("sound_id", id.toString());
		put("layer", layer.ordinal());
	}

	/**
	 * Creates a new sound event with a volume and infinite range.
	 *
	 * @param sound
	 * 		Name of sound to play.
	 * @param volume
	 * 		Volume level.
	 * @param layer
	 * 		Sound layer (e. g. ambient sound).
	 */
	public SoundEvent(final String sound, final int volume, final SoundLayer layer) {
		this(sound, layer);
		put("volume", volume);
	}

	/**
	 * Creates a new sound event using an ID with a volume and infinite range.
	 *
	 * @param id
	 * 		SoundID to send with event.
	 * @param volume
	 * 		Volume level.
	 * @param layer
	 * 		Sound layer (e. g. ambient sound).
	 */
	public SoundEvent(final SoundID id, final int volume, final SoundLayer layer) {
		this(id, layer);
		put("volume", volume);
	}

	/**
	 * Creates a new sound event with specified range and volume.
	 *
	 * @param sound
	 * 		Name of sound to play.
	 * @param radius
	 * 		Radius at which sound can be heard.
	 * @param volume
	 * 		Volume level.
	 * @param layer
	 * 		Sound layer (e. g. ambient sound).
	 */
	public SoundEvent(final String sound, final int radius, final int volume, final SoundLayer layer) {
		this(sound, volume, layer);
		put("radius", radius);
	}

	/**
	 * Creates a new sound event using an ID with specified range and volume.
	 *
	 * @param id
	 * 		SoundID to send with event.
	 * @param radius
	 * 		Radius at which sound can be heard.
	 * @param volume
	 * 		Volume level.
	 * @param layer
	 * 		Sound layer (e. g. ambient sound).
	 */
	public SoundEvent(final SoundID id, final int radius, final int volume, final SoundLayer layer) {
		this(id, volume, layer);
		put("radius", radius);
	}
}

File: src/js/stendhal/event/SoundEvent.ts
/***************************************************************************
 *                    Copyright © 2024 - Faiumoni e. V.                    *
 ***************************************************************************
 ***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU Affero General Public License as        *
 *   published by the Free Software Foundation; either version 3 of the    *
 *   License, or (at your option) any later version.                       *
 *                                                                         *
 ***************************************************************************/

import { RPEvent } from "marauroa"

import { SoundID } from "../data/sound/SoundID";
import { SoundManager } from "../data/sound/SoundManager";

import { marauroa } from "marauroa"
import { stendhal } from "../stendhal";


export class SoundEvent extends RPEvent {

	sound?: string;
	sound_id?: string;
	volume!: number;
	radius?: number;
	layer!: string;


	execute(entity: any) {
		if (!marauroa.me) {
			return;
		}
		let radius = SoundManager.DEFAULT_RADIUS;
		if (typeof(this["radius"]) === "number") {
			radius = this["radius"];
		}
		if (!marauroa.me.isInSoundRange(radius, entity)) {
			// too far away to hear so don't load
			return;
		}

		let volume = 1;
		// Adjust by the server specified volume, if any
		if (this.hasOwnProperty("volume")) {
			// NOTE: server uses int in range 1-100 while HTMLAudioElement uses float in range 0-1
			volume *= this["volume"] / 100;
		}

		let sound = this["sound"];
		// get sound from ID
		if (this["sound_id"]) {
			sound = SoundID[this["sound_id"]];
		}

		stendhal.sound.playLocalizedEffect(entity["_x"], entity["_y"], radius, this["layer"], sound,
				volume);
	}
}
